src/ceph-volume/ceph_volume/util/metadata.py

```diff
@@ -0,0 +1,438 @@
+    """Create a cache of device metadata
+
+    Uses worker threads as a producer/consumer pool where tasks to run are placed
+    on a global queue. The worker threads can then process the task, and update the
+    shared_data object
+
+    Raises:
+        ValueError: [description]
+
+    Returns:
+        [type]: [description]
+    """
+
+    # used by disk_api in Device objects
+    disk_field_list = [
+        'name',
+        'kname',
+        'pkname',
+        'disc-aln',
+        'disc-max',
+        'disc-zero',
+        'fstype',
+        'group',
+        'label',
+        'maj:min',
+        'mode',
+        'mountpoint',
+        'owner',
+        'partlabel',
+        'state',
+        'type',
+        'uuid',
+    ]
+
+    # used within sys_api in Device objects
+    blk_field_list = [
+        'disc-gran',
+        'hotplug',
+        'kname',
+        'log-sec',
+        'model',
+        'name',
+        'parttype',
+        'phy-sec',
+        'pkname',
+        'rev',
+        'rm',
+        'ro',
+        'rota',
+        'rq-size',
+        'sched',
+        'size',
+        'tran',
+        'type',
+        'vendor']
+
+    lsm_compatible_devices = ('sd',)
+    excluded_devs = ('rbd',)
+
+    def __init__(self, with_lsm: bool = False, worker_pool_size: Optional[int] = None):
+
+        self.disk = {}
+        self.lsblk = self.get_device_data()
+        dev_paths = [f"/dev/{dev}" for dev in self.lsblk.keys()]
+        dev_names = self.lsblk.keys()
+        lsm_devs = self.get_lsm_devices(dev_names)
+
+        self.set_worker_pool_size(worker_pool_size, dev_names)
+
+        self.start_workers()
+
+        self._holders = self.get_device_holders(dev_names)
+        self.metadata: Dict[str, Any] = {}
+
+        for dev in dev_paths:
+            # shared_data.udev[dev] = {}
+            task_queue.put((f'udev.{dev}', udevadm_property, (dev, DEVID_PROPS)))
+            task_queue.put((f'blkid.{dev}', blkid, (dev,)))
+
+            if with_lsm and dev in lsm_devs:
+                # shared_data.lsm[dev] = {}
+                task_queue.put((f'lsm[{dev}]', get_lsm, (dev,)))
+
+        # add the lvm based tasks
+        task_queue.put(('lv', get_lvm_data, ('lv',)))
+        task_queue.put(('pv', get_lvm_data, ('pv',)))
+        task_queue.put(('vg', get_lvm_data, ('vg',)))
+
+        task_queue.join()
+
+        self.stop_workers()
+
+        self.assemble_metadata()
+
+    def set_worker_pool_size(self, worker_pool_size: Optional[int], dev_names: List[str]):
+        """Set the side of the worker pool using device count and cpu as a guide
+
+        Args:
+            worker_pool_size (Optional[int]): requested pool size from the caller
+            dev_names (List[str]): List of devices that are on the host
+        """
+        if not worker_pool_size:
+            self.worker_pool_size = os.cpu_count()
+        else:
+            assert worker_pool_size <= os.cpu_count()
+            self.worker_pool_size = worker_pool_size
```

Comment:
Given (I assume) this is IO-intensive rather than core-bound, have you tested if the latency improves with multiples of cores (2x, 4x, ...)?

Reply:
tested at x2, and sometimes there is a benefit - but it not consistent. So I'd rather not "spin up" the threads without a consistent payback.